### LOAD DATA AND METADATA ###
source('../GRS - GJt Review Stress/FULL_DATASET/functions_for_genename_conversions.R')
rm(list = ls(pattern = '(.*)(test)|temp(.*)'))

input_genes <- readr::read_tsv('input_genes.tsv')
### LOAD DATA AND METADATA ###



create_query_for_ncbi <- get_query_for_ncbi_geneID_annotation(char_vec_gene_id_to_query_with = input_genes$Gene_name, char_vec_organism = input_genes$Species, chr_gene_identifier = 'Gene name')[[1]]

entrez_geneIDs <- search_for_ids_in_ncbi(str_vector_of_ids = create_query_for_ncbi)

input_genes$entrez_geneID <-
  as.character(lapply(
    X = entrez_geneIDs,
    FUN = function(x) {
      ifelse(test = length(x$ids) != 0 ,
             yes = x$ids[[1]],
             no = '')
    }
  ))

# These two LOCs are for some reason not found when querying for [Gene name]. it gives result for quering this LOC without [Gene name]. This is wierd... Better note that
input_genes$entrez_geneID[[59]] <- '51534'
input_genes$entrez_geneID[[80]] <- '80054'


readr::write_tsv(input_genes, 'input_genes_with_entrezID.tsv')





### GETTING PROMOTER SEQUENCES FOR GENES OF INTEREST ###
# BiocManager::install("BSgenome")
# BiocManager::install("BSgenome.Hsapiens.UCSC.hg19") # This is a genome sequence from UCSC I think
# BiocManager::install("TxDb.Hsapiens.UCSC.hg19.knownGene") # This is the annotation file from UCSC I think.


BS_homo <- BSgenome.Hsapiens.UCSC.hg19::BSgenome.Hsapiens.UCSC.hg19
TxDb_homo <- TxDb.Hsapiens.UCSC.hg19.knownGene::TxDb.Hsapiens.UCSC.hg19.knownGene


# transcript_coordinates_test <- GenomicFeatures::transcriptsBy(x = TxDb_homo, by="gene")[c('6927', '5833', '10280')]
transcript_coordinates <- GenomicFeatures::transcriptsBy(x = TxDb_homo, by="gene")[input_genes$entrez_geneID]

# promoter_sequence_test <- GenomicFeatures::getPromoterSeq(query = transcript_coordinates_test, subject = BS_homo, upstream = 1000, downstream = 2000)
promoter_sequence <- GenomicFeatures::getPromoterSeq(query = transcript_coordinates, subject = BS_homo, upstream = 1000, downstream = 2000)

# Unlist the list generated by getPromoterSeq to produce input for msa
# unlist_promoter_sequence_test <- Biostrings::unstrsplit(promoter_sequence_test)
unlist_promoter_sequence <- Biostrings::unstrsplit(promoter_sequence)
### GETTING PROMOTER SEQUENCES FOR GENES OF INTEREST ###


### MULTIPLE ALIGNMENT ###
# Muscle algoritm is easy-to-use due to being in msa package, it is very fast and reasonably accurate (Evaluating the Accuracy and Efficiency of Multiple Sequence Alignment Methods)
# We can gry using another R-compatible algorithm: MAFFT(L-INS-i) which perforomed better than Muscle here: Evaluating the Accuracy and Efficiency of Multiple Sequence Alignment Methods, though it seems to take a long time. Love you long time soldierboy. And msa takes reaal long to work either way. May need to paralelize it. This publication also suggest MAFFT: Assessing the efficiency of multiple sequence alignment programs. So lets try parallelize the shit.
# https://www.rdocumentation.org/packages/ape/versions/5.3/topics/as.alignment
# https://www.rdocumentation.org/packages/ips/versions/0.0.11/topics/mafft

# r packages for multiple alignment muscle, odseq, DECIPHER
# DECIPHER claims, that it is better to first translate nucleotide sequence to aminoacids and than align, but am not sure, cause for promoters, aminoacids do not make sense

# msa running on all 80 sequences givec below error:
# *** ERROR ***  MSA::GetLetter(0/1, 24/66577)=''/4294967295
# 
# Fatal error, exception caught.
# Error in msaFun(inputSeqs = inputSeqs, cluster = cluster, gapOpening = gapOpening,  : 
#                   MUSCLE finished by an unknown reason

# must try on shortened dataset

# Look for promoter alignment specifically

# Need to actually load the library for function to work
library(msa)
devtools::install_github("olafmersmann/microbenchmark")

mbm <- rbenchmark::benchmark("msa_test" = { msa_test <- msa::msa(inputSeqs = unlist_promoter_sequence_test, method = "Muscle") })






### MULTIPLE ALIGNMENT ###



### TRANSCRIPTION FACTOR BINDING SITES ANALYSIS ###
# TRANSFAC and JASPAR are databases of transcription factor binding models.
# Databases like ORegAnno have attempted to collect and curate individual experimentally proven TFBSs from the literature and sets from ChIP-Seq and other genome-wide experiments. 
# Many of these TF binding sites (TFBSs) are being deposited into species-specific databases (FlyBase, SGD, etc) and as tracks in the UCSC browser. tracks under 'Regulation' in the UCSC Genome Browser for your species of interest. UCSC has oreganno data: Data are also available through UCSC Genome Browser (e.g., hg38 -> Regulation -> ORegAnno). http://gtrd.biouml.org/ 

# https://bioconductor.org/packages/release/bioc/vignettes/TFBSTools/inst/doc/TFBSTools.html#introduction
# TFBSTools - package for promoter analysis
# TFBSTools::getMatrixSet() / getMatrixByID() - get data from JASPAR. Returns PFMMatrix
# TFBSTools::toPWM() - PFMMatrix -> PWMatrix?
# TFBSTools::searchSeq() - scans a nucleotide sequence with the pattern represented in the PWM. INPUT: DNAStringSet,DNAString; PWMatrix

# Other packages that may be helpful: motifStack



### TRANSCRIPTION FACTOR BINDING SITES ANALYSIS ###











# # ENSEMBL_MART_FUNCGEN The Ensembl Regulatory Build (Zerbino et al. 2015) contains a genome-wide set of regions that are likely to be involved in gene regulation.
# pa_biomart_ <- biomaRt::useMart("ENSEMBL_MART_ENSEMBL", dataset = "mmusculus_gene_ensembl")
# pa_biomart_funcgen <- biomaRt::useMart("ENSEMBL_MART_FUNCGEN")
# 
# 
# pa_test <- biomaRt::getSequence(id = 'Gapdh', 
#                                 type = "external_gene_name",
#                                 seqType="coding_gene_flank",
#                                 upstream=100, 
#                                 mart=pa_biomart_)
# 
# pa_test2 <- biomaRt::getSequence(id = 'Gapdh', 
#                                  type = "external_gene_name",
#                                  seqType="coding_gene_flank",
#                                  downstream = 100, 
#                                  mart=pa_biomart_)
# 
# ### BIOMART ###
# 
# biomartr::getMarts()
# 
# 
# pa_filters <- biomaRt::listFilters(mart = pa_biomart_)
# 
# 
# pa_mart <- biomaRt::useMart("ensembl", dataset="hsapiens_gene_ensembl")
# 
# pa_seq <- biomaRt::getSequence(id = "BRCA1",type = "hgnc_symbol",seqType = "peptide",mart = pa_mart)
# 
# show(pa_seq)
# 
# 
# ### BIOMART ###
# 
# pa_mart_test = biomaRt::useMart('ensembl')
# 
# biomaRt::listDatasets(pa_mart_test)
# 
# biomaRt::listMarts()
# listDatasets_pa_biomart_funcgen <- biomaRt::listDatasets(mart = pa_biomart_funcgen)
